test/object-store/sync/flx_sync.cpp

```diff
@@ -896,71 +898,247 @@ TEST_CASE("flx: client reset", "[sync][flx][app][baas][client reset]") {
+                REQUIRE(frozen_before->schema_version() != ObjectStore::NotVersioned);
+                REQUIRE(frozen_before->schema() == expected);
+                SharedRealm after = Realm::get_shared_realm(std::move(after_ref), util::Scheduler::make_default());
+                REQUIRE(after);
+                REQUIRE(after->schema() == expected);
+                // the above check is sufficent unless operator==() is changed to not care about ordering
+                // so future proof that by explicitly checking the order of properties here as well
+                REQUIRE(after->schema().size() == frozen_before->schema().size());
+                auto after_it = after->schema().find("TopLevel");
+                auto before_it = frozen_before->schema().find("TopLevel");
+                REQUIRE(after_it != after->schema().end());
+                REQUIRE(before_it != frozen_before->schema().end());
+                REQUIRE(after_it->name == before_it->name);
+                REQUIRE(after_it->persisted_properties.size() == before_it->persisted_properties.size());
+                REQUIRE(after_it->persisted_properties[1].name == "queryable_int_field");
+                REQUIRE(after_it->persisted_properties[2].name == "queryable_str_field");
+                REQUIRE(before_it->persisted_properties[1].name == "queryable_int_field");
+                REQUIRE(before_it->persisted_properties[2].name == "queryable_str_field");
+                REQUIRE(did_recover == (reset_mode == ClientResyncMode::Recover));
+                if (promise) {
                     promise->emplace_value();
-                };
+                }
+            };
+        };
+
+    SECTION("Recover: schema indexes match in before and after states") {
+        seed_realm(config_local, ResetMode::InitiateClientReset);
+        // reorder a property such that it does not match the on disk property order
+        std::vector<ObjectSchema> local_schema = schema;
+        std::swap(local_schema[0].persisted_properties[1], local_schema[0].persisted_properties[2]);
+        local_schema[0].persisted_properties.push_back(
+            {"queryable_oid_field", PropertyType::ObjectId | PropertyType::Nullable});
+        config_local.schema = local_schema;
+        config_local.sync_config->client_resync_mode = ClientResyncMode::Recover;
+        auto [promise, future] = util::make_promise_future<void>();
+        auto shared_promise = std::make_shared<decltype(promise)>(std::move(promise));
+        setup_reset_handlers_for_schema_validation(config_local, local_schema, shared_promise);
+        SharedRealm realm = Realm::get_shared_realm(config_local);
+        future.get();
+        CHECK(before_reset_count == 1);
+        CHECK(after_reset_count == 1);
+    }
 
-            SharedRealm realm = Realm::get_shared_realm(config_copy);
+    for (auto mode : {ClientResyncMode::DiscardLocal, ClientResyncMode::Recover}) {
```

Comment:
`auto mode = GENERATE(ClientResyncMode::DiscardLocal, ClientResyncMode::Recover)` might be a better way to do this?